### PR TITLE
Main login integration (Issues #75 & #82)

### PR DESCRIPTION
**Pull request by cjbreaux**
To tackle the main login pages, I implemented 'React Router'. There are now views and routes for the OECI login, forgotten password, and reset password. Some basic validations were added but more will be required to check for valid credentials. 

The "/passwordreset" route can only be reached manually atm as I believe it will be linked from an external email.

### CODE REVIEW

File: src/frontend/src/components/ForgotPassword/index.tsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    });
+  };
+
+  handleSubmit = (e: React.FormEvent) => {
+    e.preventDefault();
+    this.validateForm();
+  };
+
+  emailCheck = (email: string) => {
+    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
```
Comment (maxkwallace) [inline src/frontend/src/components/ForgotPassword/index.tsx]: Using a regex to validate an email address is notoriously fraught with difficulty, see https://stackoverflow.com/q/201323/1067145

I am not going to test your regex here. Just wanted to call this out and make sure the regex is relatively permissive and errs on the side of only rejecting emails that are definitely invalid.

File: src/frontend/src/components/ForgotPassword/index.tsx [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  };
+
+  validateForm = () => {
+    this.setState({
+      emailHasInput: this.state.email.trim().length === 0
```
Comment (maxkwallace) [inline src/frontend/src/components/ForgotPassword/index.tsx]: Should this be `!==`?

File: src/frontend/src/components/ForgotPassword/index.tsx [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    });
+    this.setState({
+      missingInputs: this.state.email.trim().length === 0
+    });
+    this.setState({
+      emailIsValid: this.emailCheck(this.state.email)
+    });
```
Comment (maxkwallace) [inline src/frontend/src/components/ForgotPassword/index.tsx]: I think you can combine these setState calls into one.

File: src/frontend/src/components/LoggedInHeader/index.tsx [hunk 4]
```diff
@@ -21,7 +22,10 @@ class LoggedInHeader extends React.Component {
               Account
             </button>
             <LogOut>
-              <button className="link hover-blue f6 f5-ns dib pa3">
+              <button
+                onClick={() => history.push('/')}
+                className="link hover-blue f6 f5-ns dib pa3"
+              >
```
Comment (maxkwallace) [inline src/frontend/src/components/LoggedInHeader/index.tsx]: If we use React Router's Link Component, won't it update the history automatically? That seems like a better solution. Let me know if there's something I'm missing as to why you didn't go that route. I haven't used React Router before, I'm just going off the docs at https://reacttraining.com/react-router/web/guides/quick-start/example-basic-routing

File: src/frontend/src/components/OECI/index.tsx [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    this.validateForm();
+  };
+
+  validateForm = () => {
+    this.setState({
+      userIdHasInput: this.state.userId.trim().length === 0
+    });
+    this.setState({
+      passwordHasInput: this.state.password.trim().length === 0
+    });
```
Comment (maxkwallace) [inline src/frontend/src/components/OECI/index.tsx]: Same comments as earlier:

- I think these should be `!==`?
- I think these setState calls can be combined into one.

File: src/frontend/src/components/OECI/index.tsx [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        missingInputs:
+          this.state.userId.trim().length === 0 ||
+          this.state.password.trim().length === 0
+      },
+      () =>
+        this.state.missingInputs === false
+          ? history.push('/recordsearch')
+          : null // Using setState(updater, callback) because setState doesn't immediately update component
+    );
+    //need validation for userId & PW
```
Comment (maxkwallace) [inline src/frontend/src/components/OECI/index.tsx]: Should be

    // Need validation for userId & password.

File: src/frontend/src/components/OECI/index.tsx [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+      {
+        missingInputs:
+          this.state.userId.trim().length === 0 ||
+          this.state.password.trim().length === 0
+      },
+      () =>
+        this.state.missingInputs === false
+          ? history.push('/recordsearch')
+          : null // Using setState(updater, callback) because setState doesn't immediately update component
+    );
```
Comment (maxkwallace) [inline src/frontend/src/components/OECI/index.tsx]: I realize the backend logic isn't in place so it's not possible to totally anticipate how this code should look, but we should not mutate global state (the history) as a side effect of validation. The history should only be updated if the login attempt was successful and the user is actually changing pages in response to that. So the logic in this callback should be moved to `handleSubmit` or removed for now.

File: src/frontend/src/components/PasswordReset/index.tsx [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    this.setState(
+      {
+        passwordIsValid: this.state.password.trim().length >= 10 // Need to implement a more robust password check.
+      },
+      () =>
+        this.state.passwordIsValid === true &&
+        this.state.missingPasswordInput === false
+          ? history.push('/')
+          : null // Using setState(updater, callback) beccause setState doesn't immediately update component
+    );
```
Comment (maxkwallace) [inline src/frontend/src/components/PasswordReset/index.tsx]: Same comment as above, this side effect (state change) should be in handleSubmit, not as a side effect of validation. Validation should validate only and do nothing else.

File: src/frontend/src/components/App/index.tsx [hunk 9]
```diff
@@ -2,23 +2,33 @@ import React from 'react';
...
 import { get } from 'lodash';
 import { AppState } from '../../redux/store';
+import { Router, Route } from 'react-router-dom';
 
+import history from '../History';
 import Footer from '../Footer';
 import LogIn from '../LogIn';
 import LoggedIn from '../LoggedIn';
 import RecordSearch from '../RecordSearch';
+import OECIlogin from '../OECI';
```
[the remark this answers is not in the document]
Comment (cjbreaux) [reply src/frontend/src/components/App/index.tsx]: Thanks for pointing this out Max! I can see why following a standard convention would be more important than having the names look "nice". I'll change those over to PascalCase.

File: src/frontend/src/components/App/index.tsx [hunk 10]
```diff
@@ -2,23 +2,33 @@ import React from 'react';
...
 class App extends React.Component {
   public render() {
     return get(this, 'props.system.loggedIn') ? (
-      <LoggedIn>
-        <RecordSearch />
-      </LoggedIn>
+      <Router history={history}>
+        <LoggedIn>
+          <Route path="/oeci" component={OECIlogin} />
+          <Route path="/recordsearch" component={RecordSearch} />
```
[the remark this answers is not in the document]
Comment (cjbreaux) [reply src/frontend/src/components/App/index.tsx]: That makes sense to me. I'll update the route paths.

File: src/frontend/src/components/ForgotPassword/index.tsx (hunk 1 above)
Comment (cjbreaux) [reply src/frontend/src/components/ForgotPassword/index.tsx]: Hmm, this is only checking if the syntax is correct not if the address is actually "valid". It's fairly permissive but won't accept addresses with the format:
`_@_@._`

File: src/frontend/src/components/LoggedInHeader/index.tsx (hunk 4 above)
Comment (cjbreaux) [reply src/frontend/src/components/LoggedInHeader/index.tsx]: I had intended to wrap the button in the Link component but was cautioned against it here:

https://stackoverflow.com/questions/42463263/wrapping-a-react-router-link-in-an-html-button

as nesting it that way is not valid HTML.

File: src/frontend/src/components/OECI/index.tsx (hunk 6 above)
Comment (cjbreaux) [reply src/frontend/src/components/OECI/index.tsx]: Right!

File: src/frontend/src/components/ForgotPassword/index.tsx (hunk 2 above)
Comment (cjbreaux) [reply src/frontend/src/components/ForgotPassword/index.tsx]: Ah, this kept tripping me up. The logic is correct but I've renamed the property "ariaInvalid" as I'm using it to set that property on the input field.

File: src/frontend/src/components/ForgotPassword/index.tsx (hunk 3 above)
Comment (cjbreaux) [reply src/frontend/src/components/ForgotPassword/index.tsx]: Indeed you can, thank you!

File: src/frontend/src/components/OECI/index.tsx (hunk 7 above)
Comment (cjbreaux) [reply src/frontend/src/components/OECI/index.tsx]: I've moved the logic to `handleSubmit` and changed it utilize `<Redirect to='#'>`

File: src/frontend/src/components/PasswordReset/index.tsx (hunk 8 above)
Comment (cjbreaux) [reply src/frontend/src/components/PasswordReset/index.tsx]: Using `<Redirect>` as above. Moved out of validation.

**Comment by maxkwallace** [comment]
@cjbreaux thanks for the updates! I'll take a look again tonight.

**Comment by cjbreaux** [comment]
@maxkwallace Thank you, your feedback is greatly appreciated!

File: src/frontend/src/components/ForgotPassword/index.tsx (hunk 1 above)
Comment (maxkwallace) [reply src/frontend/src/components/ForgotPassword/index.tsx]: Okay! Can you take the gist of what you wrote in your comment ^ and make that a comment in the code? It's hard for someone reading the code to make sense of a regex presented without any additional context.

File: src/frontend/src/components/LoggedInHeader/index.tsx (hunk 4 above)
Comment (maxkwallace) [reply src/frontend/src/components/LoggedInHeader/index.tsx]: Hmm, okay. In that case, could you leave a comment linking to that stackoverflow post and explaining your rationale as you just did?

**Comment by maxkwallace** [comment]
Also-- stop me if you're already familiar with this issue and don't need this comment-- it looks like this branch will need to be updated with the recent changes in `master` before it can be merged.

My personal preferred approach to doing this is to squash all the commits into one commit in the branch, then rebase that commit (i.e. the branch) on master, and then merge it. The possible downside is that you lose your commit history (if that's something you care about). 

If you have any questions about the process or run into any issues, feel free to let me know here or on Slack.

**Pull request merged**